src/libsync/syncengine.cpp

```diff
@@ -1245,6 +1245,11 @@ QString SyncEngine::adjustRenamedPath(const QString &original)
     while ((slashPos = original.lastIndexOf('/', slashPos - 1)) > 0) {
         QHash<QString, QString>::const_iterator it = _renamedFolders.constFind(original.left(slashPos));
         if (it != _renamedFolders.constEnd()) {
+            auto original_it = _renamedFolders.constFind(original);
+            if (original_it != _renamedFolders.constEnd() && it->startsWith(*original_it)) {
+                return original; // Issue #6694: "hirarchy inversion"
+            }
```

Comment:
I don't get this logic. Are you sure it doesn't indicate we should reverse the order of traversing paths in the `while`?

As far as I understand (haven't tested!) our `renamedFolders` will have two entries if "/Documents/Empty" is renamed to "/Empty/Documents":
```
/Documents -> /Empty/Documents
/Documents/Empty -> /Empty
```

Previously the function should have mapped "/Documents/Empty" to "/Empty/Documents/Empty" because the first parent directory rename is appied. With your change it'll be "/Empty/Documents" because the full path rename is taken instead.

There are probably corner cases now with three-level paths? Why aren't we taking the longest applicable rename path instead of the shortest?

Reply:
> Previously the function should have mapped "/Documents/Empty" to "/Empty/Documents/Empty" because the first parent directory rename is appied.

Yes, that's correct and that was the bug.

But with my change it will return the unchanged original .

Yes, there will be problem with three-level path when each level is renamed, or with cycle that involve several folder (as the FIXME test is testing).  But this fix for that is more complicated.

Reply:
Ok! Then the problem was that there was no indication that this was just fixing the most common problem in time for 2.5. Could you add something along the lines of `// band-aid to fix a common problem, there's a larger issue here with ... like in example ...`?